### PR TITLE
Add LibJpegTurboDecoder

### PR DESCRIPTION
**Pull request by staniewzki**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:

Implemented decoder interface for libjpeg-turbo, basing off `dali/image/jpeg.cc`.

### Affected modules and functionalities:

Copied libjpeg-turbo files into `dali/imgcodec/decoder/jpeg`.

### Key points relevant for the review:

`LibJpegTurboDecoderInstance::Decode`

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/imgcodec/decoders/decoder_test_helper.h [hunk 1]
```diff
@@ -53,7 +53,17 @@ class CpuDecoderTestBase : public ::testing::Test {
...
     Tensor<CPUBackend> result;
     EXPECT_TRUE(Decoder()->CanDecode(src, opts));
-    TensorShape<> shape = (roi.use_roi() ? roi.shape() : info.shape);
+    TensorShape<> shape;
+    if (roi) {
+      shape = roi.shape();
+      int ndim = shape.sample_dim();
+      if (ndim != info.shape.sample_dim()) {
+        shape.resize(ndim + 1);
+        shape[ndim] = info.shape[ndim];
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/decoder_test_helper.h:62]:
```suggestion
        shape[ndim] = info.shape[ndim];
        assert(shape.sample_dim() == info.shape.sample_dim());
```

File: dali/imgcodec/decoders/libjpeg_turbo.cc [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    flags.crop_width = roi.shape()[1];
+  }
+
+  const uint8_t *encoded_data;
+  size_t data_size;
+  if (in->Kind() == InputKind::HostMemory) {
+    encoded_data = in->RawData<uint8_t>();
+    data_size = in->Size();
+  } else {
+    DALI_FAIL("InputKind not supported");
```
Comment (jantonguirao) [inline dali/imgcodec/decoders/libjpeg_turbo.cc:58]: I think it'd be good to print the input kind received (the integer representation would be enough)

File: dali/imgcodec/decoders/libjpeg_turbo_test.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+template<typename OutputType>
+class LibJpegTurboDecoderTest : public NumpyDecoderTestBase<OutputType> {
+ protected:
+  static const auto dtype = type2id<OutputType>::value;
+
+  std::shared_ptr<ImageDecoderInstance> CreateDecoder(ThreadPool &tp) override {
+    return LibJpegTurboDecoder{}.Create(CPU_ONLY_DEVICE_ID, tp);
+  }
+
+  std::shared_ptr<ImageParser> CreateParser() override {
```
[the remark this answers is not in the document]
Comment (staniewzki) [reply dali/imgcodec/decoders/libjpeg_turbo_test.cc:80]: Not used, but CreateParser is an abstract method, so there must be some implementation provided.

**Review by jantonguirao — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**